src/mlpack/core/optimizers/sgdr/sgdr_impl.hpp

```diff
@@ -0,0 +1,193 @@
+                            batchSize,
+                            stepSize,
+                            maxIterations,
+                            tolerance,
+                            shuffle,
+                            updatePolicy,
+                            SnapshotEnsembles(
+                                epochRestart,
+                                multFactor,
+                                stepSize,
+                                batchSize,
+                                function.NumFunctions(),
+                                maxIterations,
+                                snapshots)))
+{
+  /* Nothing to do here */
+}
+
+template<
+    typename DecomposableFunctionType,
+    typename UpdatePolicyType,
+    typename DecayPolicyType
+>
+template<typename PolicyType>
+SGDR<
+    DecomposableFunctionType,
+    UpdatePolicyType,
+    DecayPolicyType
+>::SGDR(DecomposableFunctionType& function,
+        const size_t epochRestart,
+        const double multFactor,
+        const size_t batchSize,
+        const double stepSize,
+        const size_t maxIterations,
+        const double tolerance,
+        const bool shuffle,
+        const UpdatePolicyType& updatePolicy,
+        const typename std::enable_if_t<std::is_same<
+            PolicyType, CyclicalDecay>::value>* /* junk */) :
+    function(function),
+    batchSize(batchSize),
+    accumulate(true),
+    optimizer(OptimizerType(function,
+                            batchSize,
+                            stepSize,
+                            maxIterations,
+                            tolerance,
+                            shuffle,
+                            updatePolicy,
+                            CyclicalDecay(
+                                epochRestart,
+                                multFactor,
+                                stepSize,
+                                batchSize,
+                                function.NumFunctions())))
+{
+  /* Nothing to do here */
+}
+
+template<
+    typename DecomposableFunctionType,
+    typename UpdatePolicyType,
+    typename DecayPolicyType
+>
+template<typename PolicyType>
+double SGDR<
+    DecomposableFunctionType,
+    UpdatePolicyType,
+    DecayPolicyType
+>::Optimize(arma::mat& iterate,
+            const typename std::enable_if_t<std::is_same<
+                PolicyType, SnapshotEnsembles>::value>* /* junk */)
+{
+  // If a user changed the step size he hasn't update the step size of the
+  // cyclical decay instantiation, so we have to do here.
+  if (optimizer.StepSize() != optimizer.DecayPolicy().StepSize())
+  {
+    optimizer.DecayPolicy().StepSize() = optimizer.StepSize();
+  }
+
+  // If a user changed the batch size we have to update the restart fraction
+  // of the cyclical decay instantiation.
+  if (optimizer.BatchSize() != batchSize)
+  {
+    batchSize = optimizer.BatchSize();
+    optimizer.DecayPolicy().EpochBatches() = function.NumFunctions() /
+        double(batchSize);
+  }
```

Comment:
One possibility here might be to add a function to `DecayPolicy()` called something like `UpdateParameters(OptimizerType&)`, which for `CyclicalDecay` or `SnapshotEnsembles` could do the same thing as this code and for `NoDecay` could do nothing.  This could help reduce the need for two separate implementations here.